crates/swc_ecma_lints/src/rules/no_loop_func.rs

```diff
@@ -0,0 +1,305 @@
+                names.join(", ")
+            )
+        } else {
+            names.sort_unstable();
+
+            format!(
+                "Function declared in a loop contains unsafe references to variable(s) {}",
+                names.join(", ")
+            )
+        };
+
+        HANDLER.with(|handler| match self.expected_reaction {
+            LintRuleReaction::Error => {
+                handler.struct_span_err(span, &message).emit();
+            }
+            LintRuleReaction::Warning => {
+                handler.struct_span_warn(span, &message).emit();
+            }
+            _ => {}
+        });
+    }
+
+    fn is_unsafe_variable(&self, id: &Id) -> bool {
+        self.scopes
+            .iter()
+            .rev()
+            .any(|scope| self.scoped_unsafe_vars.get(scope).unwrap().contains(id))
+    }
+
+    fn extact_vars(&mut self, pat: &Pat) {
+        match pat {
+            Pat::Ident(ident) => {
+                self.scoped_unsafe_vars
+                    .get_mut(self.scopes.last().unwrap())
+                    .unwrap()
+                    .insert(ident.to_id());
+            }
+            Pat::Array(ArrayPat { elems, .. }) => {
+                elems.iter().for_each(|elem| {
+                    if let Some(elem) = elem {
+                        self.extact_vars(elem);
+                    }
+                });
+            }
+            Pat::Object(ObjectPat { props, .. }) => {
+                props.iter().for_each(|prop| match prop {
+                    ObjectPatProp::Assign(AssignPatProp { key, .. }) => {
+                        self.scoped_unsafe_vars
+                            .get_mut(self.scopes.last().unwrap())
+                            .unwrap()
+                            .insert(key.to_id());
+                    }
+                    ObjectPatProp::KeyValue(KeyValuePatProp { value, .. }) => {
+                        self.extact_vars(value.as_ref());
+                    }
+                    ObjectPatProp::Rest(RestPat { arg, .. }) => {
+                        self.extact_vars(arg.as_ref());
+                    }
+                });
+            }
+            Pat::Rest(RestPat { arg, .. }) => {
+                self.extact_vars(arg.as_ref());
+            }
+            Pat::Assign(AssignPat { left, .. }) => {
+                self.extact_vars(left.as_ref());
+            }
+            Pat::Invalid(_) => {}
+            Pat::Expr(_) => {}
+        }
+    }
+}
+
+impl Visit for NoLoopFunc {
+    noop_visit_type!();
+
+    fn visit_block_stmt(&mut self, block: &BlockStmt) {
+        self.scopes.push(block.span);
+        self.scoped_unsafe_vars
+            .insert(block.span, Default::default());
+
+        block.visit_children_with(self);
+
+        self.scopes.pop();
+        self.scoped_unsafe_vars.remove(&block.span);
+    }
+
+    fn visit_for_stmt(&mut self, for_stmt: &ForStmt) {
+        self.loop_depth += 1;
+
+        self.inside_loop_decl = true;
```

Comment:
I think this value can be wrong for nested loops.

```js
for (;;) {
   for (;;);

   // At here, `inside_loop_decl` will be false.
   foo()
}
```

Reply:
I think not. Firstly I check loop decl (1)

```
    fn visit_for_in_stmt(&mut self, for_in_stmt: &ForInStmt) {
        self.loop_depth += 1;

        self.inside_loop_decl = true;  // start of checking loop head

        for_in_stmt.left.visit_children_with(self);        // check head
        for_in_stmt.right.visit_children_with(self);

        self.inside_loop_decl = false; // end of checking loop head

        for_in_stmt.body.visit_children_with(self); // check body

        self.loop_depth -= 1;
    }

```

Reply:
Oh, then the problem is

```js
for (var a = () => {
  for (;;);
   // At here, `inside_loop_decl` will be false.
};;) {
   
}
```

but I don't expect people to write this code.
So it's fine I guess?

Reply:
I think yes